src/entities/trade.test.ts

```diff
@@ -0,0 +1,661 @@
+      const route2v3 = new RouteV3(route2OriginalV3)
+      const amount2v3 = CurrencyAmount.fromRawAmount(token0, JSBI.BigInt(1000))
+
+      await expect(
+        Trade.fromRoutes(
+          [{ routev2, amount: amountv2 }],
+          [
+            { routev3, amount: amountv3 },
+            { routev3: route2v3, amount: amount2v3 },
+          ],
+          TradeType.EXACT_INPUT
+        )
+      ).rejects.toThrow('POOLS_DUPLICATED')
+    })
+
+    it('throws if pools are re-used between V2 routes', async () => {
+      const routeOriginalV2 = new V2RouteSDK([pair_0_1, pair_1_2], token0, token2)
+      const routev2 = new RouteV2(routeOriginalV2)
+      const amountv2 = CurrencyAmount.fromRawAmount(token0, JSBI.BigInt(100))
+
+      const route2OriginalV2 = new V2RouteSDK([pair_0_1, pair_weth_1, pair_weth_2], token0, token2)
+      const route2v2 = new RouteV2(route2OriginalV2)
+      const amount2v2 = CurrencyAmount.fromRawAmount(token0, JSBI.BigInt(1000))
+
+      const routeOriginalV3 = new V3RouteSDK([pool_0_1, pool_1_2], token0, token2)
+      const routev3 = new RouteV3(routeOriginalV3)
+      const amountv3 = CurrencyAmount.fromRawAmount(token0, JSBI.BigInt(1000))
+
+      await expect(
+        Trade.fromRoutes(
+          [
+            { routev2, amount: amountv2 },
+            { routev2: route2v2, amount: amount2v2 },
+          ],
+          [{ routev3, amount: amountv3 }],
+          TradeType.EXACT_INPUT
+        )
+      ).rejects.toThrow('POOLS_DUPLICATED')
+    })
+
+    it('throws if routes have different inputs', async () => {
+      const routeOriginalV2 = new V2RouteSDK([pair_1_2], token1, token2)
+      const routev2 = new RouteV2(routeOriginalV2)
+      const amountv2 = CurrencyAmount.fromRawAmount(token1, JSBI.BigInt(100))
+
+      const routeOriginalV3 = new V3RouteSDK([pool_0_1, pool_1_2], token0, token2)
+      const routev3 = new RouteV3(routeOriginalV3)
+      const amountv3 = CurrencyAmount.fromRawAmount(token0, JSBI.BigInt(1000))
+
+      await expect(
+        Trade.fromRoutes([{ routev2, amount: amountv2 }], [{ routev3, amount: amountv3 }], TradeType.EXACT_INPUT)
+      ).rejects.toThrow('INPUT_CURRENCY_MATCH')
+    })
+
+    it('throws if routes have different outputs', async () => {
+      const routeOriginalV2 = new V2RouteSDK([pair_0_1], token0, token1)
+      const routev2 = new RouteV2(routeOriginalV2)
+      const amountv2 = CurrencyAmount.fromRawAmount(token0, JSBI.BigInt(100))
+
+      const routeOriginalV3 = new V3RouteSDK([pool_0_1, pool_1_2], token0, token2)
+      const routev3 = new RouteV3(routeOriginalV3)
+      const amountv3 = CurrencyAmount.fromRawAmount(token0, JSBI.BigInt(1000))
+
+      await expect(
+        Trade.fromRoutes([{ routev2, amount: amountv2 }], [{ routev3, amount: amountv3 }], TradeType.EXACT_INPUT)
+      ).rejects.toThrow('OUTPUT_CURRENCY_MATCH')
+    })
+  })
+  describe('#worstExecutionPrice', () => {
```

Comment:
Can we test #minimumAmountIn / #maximumAmountOut / #priceImpact / #executionPrice too?